### PR TITLE
gluetool-html-log: from JSON log, generate nice HTML log

### PR DESCRIPTION
**Pull request by happz**

### CODE REVIEW

**Comment by happz** [comment]
I broke `bash-completion` module to raise an exception, catch it & raise another one. That chains both exceptions together, mentioning them both in the log. Then, when I fed `gluetool-html-log` with the JSON log (captured via `--json-file` option), I get a nice HTML page :)

**Complete log** - note missing `DEBUG` options, it is possible to generate log with all log entries by setting `-D` option of `gluetool-html-log`:
log

**Exceptions** - when one clicks on "Exceptions" button, list of exceptions and their properties opens, listing all their frames:
exceptions

It is possible to open **code snippet** with highlighting line:
code

And, of course, **local variables** of each frame:
locals

In the future, I'd like to add some sort of filtering form on top of the page (or on the right side, in a floating window maybe), to allow showing just entries from a given thread, with given substring, produced by a given module and so on.

File: gluetool/html_log.py [hunk 1]
```diff
@@ -0,0 +1,424 @@
...
+    if ctx.autoescape:
+        return Markup(result)
+
+    return result
+
+
+def _snippet_filter(ctx, filepath, lineno, syntax, window=10):
+    # type: (Any, str, int, str, int) -> Union[str, Markup]
+    """
+    Generic filter providing snippet of the code from a source file, including the highlight.
```
Comment (mkluson) [inline gluetool/html_log.py:275]: Why are you not mention `ctx` in docstrings?

**Review by mkluson — DISMISSED** [review]

**Review by optak — DISMISSED** [review]

**Review by thrix — DISMISSED** [review]
Nice work @happz, tried it and a nice start of html logging. No issues really spotted.

File: gluetool/html_log.py (hunk 1 above)
Comment (happz) [reply gluetool/html_log.py:275]: It's a context provided by Jinja, but you're right, it should be mentioned. Added in the recent patch.

**Pull request merged**